src/main/java/org/truffleruby/language/objects/WarnTopLevelClassVariableAccessNode.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.language.objects;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.source.SourceSection;
+import org.truffleruby.language.RubyBaseNode;
+import org.truffleruby.language.WarnNode;
+
+public class WarnTopLevelClassVariableAccessNode extends RubyBaseNode {
+
+    @Child private WarnNode warnNode = new WarnNode();
+
+    @TruffleBoundary
+    public void warnTopLevelClassVariableAccess(SourceSection sourceSection) {
+        if (sourceSection != null) {
+            warnNode.warn(sourceSection.getSource().getName(), ":", Integer.toString(sourceSection.getStartLine()),
+                    ": warning: class variable access from toplevel");
+        } else {
+            warnNode.warn("warning: class variable access from toplevel");
+        }
```

Comment:
This makes me think `WarnNode` should handle adding the source info itself, and require a SourceSection (or null) to be passed. Can be another PR.

Reply:
The `warning: ` part can also likely be shared.

Reply:
I've refactored, please re-review.

Reply:
Nice :)